File: Linear/Dynamic/Linked List/Doubly Linked List/03_Deleting/Given_Position/Deleting.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

def deleteGivenNode(head, pos):
    temp = head

    if head is None or pos < 1:
        return None
    
    # Head is to be deleted
    if pos is 1:
        head = head.next
        temp = None
        if head is not None:
            head.prev = None
        return head
    
    # Traverse to the node before the one to be deleted
    for i in range(1, pos, 1):
        if temp is None:
            return head
        temp = temp.next
    
    if temp is None:
        return head
    
    if temp.prev is not None:
        temp.prev.next = temp.next
    
    if temp.next is not None:
        temp.next.prev = temp.prev
    
    return head

def printList(head):
    while head is not None:
        print(head.data, end="")
        if head.next is not None:
            print(" <-> ", end="")
        head = head.next
    print()

# Create a hardcoded doubly linked list:
# 1 <-> 2 <-> 3
head = Node(1)
second = Node(2)
third = Node(3)

head.next = second
second.prev = head
second.next = third
third.prev = second

print("Linked List before deletion\n")
printList(head)

pos = int(input("Enter a position: \n"))
head = deleteGivenNode(head, pos)

print("Linked List after deletion\n")
printList(head)

File: Linear/Dynamic/Linked List/Doubly Linked List/03_Deleting/Given_Position/Deleting.js
import readline from 'readline'

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
})

class Node {
    constructor(data) {
        this.data = data;
        this.prev = null;
        this.next = null;
    }
}

function deleteGivenNode(head, pos) {
    let temp = head;

    if (head === null || pos < 1)
        return null;

    // Head is to be deleted
    if (pos === 1) {
        head = head.next;
        temp = null;
        if (head !== null)
            head.prev = null;
        return head;
    }

    // Traverse to the node before the one to be deleted
    for (let i = 1; i < pos && temp !== null; i++) {
        temp = temp.next;
    }

    if (temp === null)
        return head;

    if (temp.prev !== null)
        temp.prev.next = temp.next;

    if (temp.next !== null)
        temp.next.prev = temp.prev;

    return head;
}

function printList(head) {
    while (head !== null) {
        process.stdout.write(head.data.toString())
        if (head.next !== null)
            process.stdout.write(" <-> ")
        head = head.next
    }
    console.log()
}

// Create a hardcoded doubly linked list:
// 1 <-> 2 <-> 3
let head = new Node(1);
let second = new Node(2);
let third = new Node(3);

head.next = second;
second.prev = head;
second.next = third;
third.prev = second;

console.log("Linked List before deletion");
printList(head);

rl.question("Enter a position: ", (ans) => {
    const pos = parseInt(ans)

    head = deleteGivenNode(head, pos);

    console.log("Linked List after deletion");
    printList(head);
    
    rl.close()
})